#166151257 Integrate with Fitbit

willeswa opened the pull request:
#### What does this PR do?
It integrates Fitbit API
#### Description of Task to be completed?
- [x] Configures a FitBit application
- [x] Implement OAuth2 for the implementation
#### How should this be manually tested?
- [x] Create a Fitbit account: set the redirect uri, get the client_id and secret and set them in your env
- [x] Clone the repo and set it up for development.
- [x] Checkout out to the `ft-fitbit-support-166151257` branch & create an account
- [x] On your Wger dashboard, you should be able to see your name on the upper right corner
- [x] Click on the dropdown and click on Sync Fitbit
- [x] You should be able to authorize and be redirected back to the dashboard
#### Any background context:
When creating a testing application on fitbit, the call backurl should point to this endpoint ```/en/user/sync_fitbit/```
#### Story ID
[166151257](https://www.pivotaltracker.com/story/show/166151257)
#### Screenshots
<img width="1433" alt="Screen Shot 2019-06-13 at 22 52 12" src="https://user-images.githubusercontent.com/38868860/59464815-1dbdd580-8e32-11e9-9a02-7f094e689d7e.png">
<img width="1422" alt="Screen Shot 2019-06-13 at 22 52 24" src="https://user-images.githubusercontent.com/38868860/59464816-1dbdd580-8e32-11e9-98a8-1fbcd0914c93.png">
<img width="1323" alt="Screen Shot 2019-06-13 at 22 52 41" src="https://user-images.githubusercontent.com/38868860/59464817-1dbdd580-8e32-11e9-9545-9d99996de161.png">
<img width="1425" alt="Screen Shot 2019-06-13 at 22 53 09" src="https://user-images.githubusercontent.com/38868860/59464820-1eef0280-8e32-11e9-8f73-73f0d89248f7.png">
<img width="1119" alt="Screen Shot 2019-06-13 at 23 07 27" src="https://user-images.githubusercontent.com/38868860/59464821-1eef0280-8e32-11e9-8fce-c04db771261a.png">
<img width="872" alt="Screen Shot 2019-06-13 at 23 23 24" src="https://user-images.githubusercontent.com/38868860/59464882-49d95680-8e32-11e9-9ca5-db0d97c6e855.png">

wangonya commented on wger/core/views/user.py:
```diff
@@ -40,6 +43,11 @@
 
+from fitbit import FitbitOauth2Client
+import base64
+import requests
```
You can move the import statements here to the top.

ceexon replied:
Why should he make the change yet flake 8 has not raised any complaints.

wangonya replied:
This is just to conform to pep8 standards. According to [this](https://pep8.org/#imports) resource, import statements should be grouped as follows:

1. standard library imports
2. related third party imports
3. local application/library specific imports

He has them mixed up here.

willeswa committed:
166151257-feat(api intergration): intergrate with fitbit

- fitbit should be able to connect to fitbit
- app fetch data from fitbit

[Delivers #166151257]

wangonya reviewed (commented):
Your PR description message seems to be a bit different. Any reason you chose to go with this format?
Also, the PR heading does not conform to standards. Kindly refer to https://github.com/andela/bestpractices/wiki/Git-naming-conventions-and-best-practices

willeswa commented:
> Your PR description message seems to be a bit different. Any reason you chose to go with this format?
> Also, the PR heading does not conform to standards. Kindly refer to https://github.com/andela/bestpractices/wiki/Git-naming-conventions-and-best-practices

The PR format follows this convention introduced during the Deferral program. Refer to https://docs.google.com/document/d/1k1vITs3zDuVS7l2gCY5ML-kqAOq7euQIBlWrNonwhYw/edit?usp=sharing. I just realized, the above were team conventions. I have updated the PR.

jmusila reviewed (commented):
I found the PR description a little bit hard to follow when testing locally, maybe you can add more descriptions or if there are environment variables to be exported, you can add them in the `.env_sample` file, that is, if there are more variables.

wekesa931 commented on wger/core/views/user.py:
```diff
@@ -323,6 +330,82 @@ def preferences(request):
+
+    """
+
+    if settings.WGER_SETTINGS['FITBIT_CLIENT_ID'] is None:
```
Impressive work! 🙌🏼 You could consider testing for the availability of a secret key too.

willeswa replied:
> Impressive work! 🙌🏼 You could consider testing for the availability of a secret key too.

Working on it

wekesa931 commented on wger/core/views/user.py:
```diff
@@ -323,6 +330,82 @@ def preferences(request):
+            date = (datetime.date.today()).strftime('%Y-%m-%d')
+
+            response = requests.get(
+                'https://api.fitbit.com/1/user/-/body/log/weight/date/' + date + '.json',
```
Is this uri hardcoded or is it customized for every user that logs into wger?

willeswa replied:
> Is this uri hardcoded or is it customized for every user that logs into wger?

It varies for each user. I think a lot has been abstracted by Fitbit. It seems that once you authorize Wger to get Fitbit data, that uri will somehow be custom to your account only.

willeswa commented:
> I found the PR description a little bit hard to follow when testing locally, maybe you can add more descriptions or if there are environment variables to be exported, you can add them in the `.env_sample` file, that is, if there are more variables.

Thank you for pointing out this. I have updated the .sample.env. Does it make it easier now?

wangonya commented on wger/core/views/user.py:
```diff
@@ -323,6 +330,82 @@ def preferences(request):
+@login_required
+def sync_fitbit(request):
+    """Authorizes the connection with Wger
+
```
The formatting here is a bit off. You can either bring the lower triple quotes up to line 335 or take the `Authorizes the connection with Wger` down to line 336.

willeswa replied:
> The formatting here is a bit off. You can either bring the lower triple quotes up to line 335 or take the `Authorizes the connection with Wger` down to line 336.

Noted.

caveinn commented on .env_sample:
```diff
@@ -9,4 +9,7 @@ export SECRET_KEY=''
+export RECAPTCHA_PRIVATE_KEY=''
+export FITBIT_CLIENT_ID=22DRQ3
+export FITBIT_CLIENT_SECRET=a9c286f26f3228fc572a53b93bea5b73
+export ROOT_URL='http://127.0.0.1:8000'
```
It is good practice to have sensitive information such as the keys not visible on the codebase. I would advise having this on the manually configurable env settings on both travis and heroku.

willeswa replied:
This is a sample .env and the secret codes are not real or valid. They are supposed to show what is to be added to the real .env

caveinn reviewed (commented):
Awesome work. Just check on the comment I left.

caveinn reviewed (approved):
LGTM
